File: php/src/Advertisements/User/Application/Command/SignUpMember/SignUpMemberUseCase.php
<?php
declare(strict_types=1);

namespace Demo\App\Advertisements\User\Application\Command\SignUpMember;

use Demo\App\Advertisements\Shared\ValueObjects\CivicCenterId;
use Demo\App\Advertisements\Shared\ValueObjects\Email;
use Demo\App\Advertisements\Shared\ValueObjects\Password;
use Demo\App\Advertisements\Shared\ValueObjects\UserId;
use Demo\App\Advertisements\User\Domain\Exceptions\AdminWithIncorrectCivicCenterException;
use Demo\App\Advertisements\User\Domain\Exceptions\MemberAlreadyExistsException;
use Demo\App\Advertisements\User\Domain\Exceptions\UserNotFoundException;
use Demo\App\Advertisements\User\Domain\MemberUser;
use Demo\App\Advertisements\User\Domain\UserRepository;
use Demo\App\Advertisements\User\Domain\ValueObjects\MemberNumber;
use Demo\App\Advertisements\User\Domain\ValueObjects\Role;
use Exception;

final class SignUpMemberUseCase
{
    public function __construct(private UserRepository $userRepository)
    {
    }

    /**
     * @throws Exception
     */
    public function execute(SignUpMemberCommand $command): void
    {
        // TODO: Use security service
        $adminUser = $this->userRepository->findAdminById(new UserId($command->securityUserId));
        if (!$adminUser) {
            throw UserNotFoundException::asAdmin();
        }

        if (!$adminUser->civicCenterId()->equals(new CivicCenterId($command->civicCenterId))) {
            throw AdminWithIncorrectCivicCenterException::differentCivicCenterFromMember();
        }

        if ($this->userRepository->findMemberByIdOrNull(new UserId($command->memberId))) {
            throw MemberAlreadyExistsException::build();
        }

        $member = MemberUser::signUp(
            new UserId($command->memberId),
            new Email($command->email),
            Password::fromPlainPassword($command->password),
            Role::MEMBER,
            new MemberNumber($command->memberNumber),
            new CivicCenterId($command->civicCenterId),
        );

        $this->userRepository->saveMember($member);
    }
}

File: kotlin/src/main/kotlin/advertisements/user/application/command/signupmember/SignUpMemberUseCase.kt
package advertisements.user.application.command.signupmember

import advertisements.shared.value_object.CivicCenterId
import advertisements.shared.value_object.Email
import advertisements.shared.value_object.Password
import advertisements.shared.value_object.UserId
import advertisements.user.domain.MemberUser
import advertisements.user.domain.UserRepository
import advertisements.user.domain.exceptions.AdminWithIncorrectCivicCenterException
import advertisements.user.domain.exceptions.MemberAlreadyExistsException
import advertisements.user.domain.exceptions.UserNotFoundException
import advertisements.user.domain.value_object.MemberNumber
import advertisements.user.domain.value_object.Role
import framework.database.TransactionManager

class SignUpMemberUseCase(
    private val userRepository: UserRepository,
    private val transactionManager: TransactionManager
) {

    fun execute(command: SignUpMemberCommand) {
        transactionManager.beginTransaction()
        try {
            val adminUser = userRepository.findAdminById(UserId(command.securityUserId))
                ?: throw UserNotFoundException.asAdmin()

            if (!adminUser.civicCenterId().equals(CivicCenterId.create(command.civicCenterId))) {
                throw AdminWithIncorrectCivicCenterException.differentCivicCenterFromMember()
            }

            if (userRepository.findMemberById(UserId(command.memberId)) != null) {
                throw MemberAlreadyExistsException.build()
            }

            val member = MemberUser.signUp(
                UserId(command.memberId),
                Email(command.email),
                Password.fromPlainPassword(command.password),
                Role.MEMBER,
                MemberNumber(command.memberNumber),
                CivicCenterId.create(command.civicCenterId)
            )

            userRepository.saveMember(member)
            transactionManager.commit()
        } catch (e: Exception) {
            transactionManager.rollback()
            throw e
        }
    }
}
